torch/_functorch/aot_autograd.py

```diff
@@ -1647,6 +1587,106 @@ def remove_dupe_metadata(
+    synthetic_base_info: List[Union[int, Tuple[int, torch.Tensor]]],
+    flat_args: List[Any],
+    flat_args_with_synthetic_bases: List[Any],
+) -> Tuple[ViewAndMutationMeta, List[int]]:
+
+    # For each arg in the inner calling convention, map it to its idx in the outer calling convention
+    # (there might be multiple indices, if the inner arg is a synthetic base)
+    synthetic_base_to_indices: List[List[int]] = []
+    for idx in range(len(flat_args_with_synthetic_bases)):
+        outer_aliased_indices_of_current_base_arg = [
+            i for i, outer_idx_or_tuple in enumerate(synthetic_base_info)
+            if (isinstance(outer_idx_or_tuple, int) and outer_idx_or_tuple == idx)
+            or (isinstance(outer_idx_or_tuple, tuple) and outer_idx_or_tuple[0] == idx)
+        ]
+        synthetic_base_to_indices.append(outer_aliased_indices_of_current_base_arg)
+
+    # given the requires_grad info on mutated inputs,
+    # generate the requires_grad info on those same mutated inputs, but after constructing synthetic bases.
+    input_infos = []
+    mutated_inp_require_grad_info = []
+    for i, indices in enumerate(synthetic_base_to_indices):
+        inpt_info = InputAliasInfo(
+            # If len(indices) > 1, then this input is a synthetic base.
+            # The invariant is that to the rest of aot autograd, synthetic bases only show up if
+            # one of their aliases gets a data mutation. And if any of their aliases get metadata
+            # mutations, they will be hidden from the rest of aot autograd.
+            mutates_data=True if len(indices) > 1 else m.input_info[indices[0]].mutates_data,
+            mutates_metadata=False if len(indices) > 1 else m.input_info[indices[0]].mutates_metadata,
+            # leaf-ness should be all-or-nothing for aliased tensor.
+            is_leaf=any(m.input_info[x].is_leaf for x in indices),
+        )
+        input_infos.append(inpt_info)
+        # requires_grad_info consists of (mutated_inputs, forward_outputs).
+        # For any mutated inputs that correspond to aliased inputs,
+        # Need to replace them with their mutated synthetic base
+        if inpt_info.mutates_data or inpt_info.mutates_metadata:
+            mutated_inp_require_grad_info.append(any(m.requires_grad_info[x] for x in indices))
+
+    # Find any inputs that fulfill the following criteria:
+    # (1) They are part of a synthetic base (because they alias another input,
+    #      and at least one input experiences a data mutation)
+    # (2) They experience a metadata mutation
+    aliased_arg_idx_with_metadata_mutations = [
+        i for i, inpt_info in enumerate(m.input_info)
+        if inpt_info.mutates_metadata and not isinstance(synthetic_base_info[i], int)
+    ]
+
+    # grab the original requires grad info on the outputs, except the ones from the mutated inputs
+    num_original_input_data_mutations = len([x for x in m.input_info if x.mutates_data or x.mutates_metadata])
```

Comment:
yeah I would really really love if we could assume there's no metadata mutation at this point anymore

Reply:
for a future refactor :) (discussed offline and I think we agreed this would be nice, but not necessary for now)